### ISSUE TITLE
Add type hinting to the Python module

### ISSUE THREAD

**Issue by bryanwweber** [Cantera/enhancements#85]
**Abstract**

This feature request will add Python [PEP-484](https://www.python.org/dev/peps/pep-0484/) compliant type hints to the Python module.

**Motivation**

The use of static type hinting in Python is increasingly recognized as best practice, and many IDE and text editor tools (for example, PyCharm and VS Code + Python extension) ship with support for static type checking. This improves correctness of code, especially for larger scripts and projects. In addition, type hinting is entirely opt-in and does not affect runtime behavior in any way.

**Possible Solutions**

There are two possible solutions. One is to use the existing Cython types (where they exist) and add the missing ones in the existing Python module. I'm not sure if this is possible or not.

Alternatively, we can ship (either as part of the main Cantera package, or in a separate "stubs" package) a stub file that describes the interface. Ideally, this file could be partly autogenerated during build, most likely from a template.

**References**

https://mypy.readthedocs.io/en/stable/installed_packages.html#making-pep-561-compatible-packages
https://mypy.readthedocs.io/en/stable/getting_started.html#stubs-intro
https://mypy.readthedocs.io/en/stable/stubs.html

**Comment by ischoegl** [comment]
https://cython.readthedocs.io/en/latest/src/tutorial/pure.html#pep-484-type-annotations

**Comment by TimothyEDawson** [comment]
I'd like to resurrect this feature request, as Cantera type hints would be extremely useful for some libraries I am developing. Type hints have become the de facto standard for modern Python packages, and are highly recommended by relevant entities such as the [Scientific Python project](https://scientific-python.org/).

It looks to me like creating an independent cantera-stubs package would be the best path forward. That would allow some flexibility and experimentation, while still maintaining the possibility of bringing static typing into the core codebase once the stubs package reaches a stable state. I would be tentatively interested in leading this effort, but first I wanted to check to see if any work has already been done to this effect?

**Comment by ischoegl** [comment]
> It looks to me like creating an independent cantera-stubs package would be the best path forward. That would allow some flexibility and experimentation, while still maintaining the possibility of bringing static typing into the core codebase once the stubs package reaches a stable state. I would be tentatively interested in leading this effort, but first I wanted to check to see if any work has already been done to this effect?

Thank you for the inquiry, @TimothyEDawson! To my knowledge, there isn't a current and/or ongoing effort. Any contributions would be welcome!

Regarding an independent cantera-stubs package, I'm not sure which is more effort, as things would have to be merged eventually. One open question is to find a good approach to [Cython static typing](https://docs.cython.org/en/stable/src/tutorial/pure.html). As most of Cantera's Python API originates from Cython, this goes hand-in-hand with PEP 484. For an effort like this, it may be useful to select a small module and provide a proof of concept PR so details can be hammered out.

**Comment by TimothyEDawson** [comment]
They wouldn't necessarily need to ever be merged, but an independent stubs package might provide a good reference for an eventual core implementation. At which point the stubs package would be retired, having served its purpose as a stopgap.

I've had success in my personal projects creating a minimal set of Cantera stubs wherein I basically treat Cantera as if it were written in pure Python. I've found this works quite well for the type of libraries which would interface with Cantera through its Python API, as it seems to be extremely rare that external code would need to know anything about what's happening at the Cython level.

Knowing my personal capabilities and needs, I may try and create such a "simple" stubs package and see where that goes. My goal with it would be to get good coverage of the Python API in order to support use of Cantera within typed libraries (think: broad and shallow), rather than dive deep into Cython static typing with a proof of concept (narrow and deep) as you suggest. I like the idea, though, and hope to be able to contribute to such an effort in the near future!

**Comment by ischoegl** [comment]
@TimothyEDawson ... thanks for providing more context. I looked some more into `pyi` files and see that they're typically independent of a main library, so your suggestion is workable. At the same time, I do have a concern that a separate package will create additional maintenance burden long-term (presumably, that library would still be part of the Cantera project). I'll let some of the other team members chime in.

It would, however, be useful to learn more about your use case for type hinting?

**Comment by speth** [comment]
I agree that there are likely to be a few issues with providing the type hints directly in the Cython module. Regarding the option of using stubs, my suggestion would be to simply include the stub files in the main Cantera repository and Python package, rather than creating a new package that needs to be maintained separately. Even an incomplete set of type hints would be a step in the right direction.

**Comment by TimothyEDawson** [comment]
@ischoegl Mainly it's to get my static type checkers (Mypy, basedpyright, pyrefly, etc.) to stop complaining about every piece of code which touches a Cantera function, without using workarounds such as sticking `#ignore` directives everywhere. But it is also quite helpful for various IDE features.

@speth I was envisioning this as beginning life as a third-party, completely optional stub library, which would bring with it no official guarantees and require no official support. This could then be brought into the Cantera ecosystem after proving its utility. But I suppose if there's interest in using and maintaining type hints, I would be interested in going straight to working on a pull request for direct implementation, and see where that goes.

While we could certainly place .pyi stub files alongside the Cython code, I suspect it wouldn't actually be too difficult to put them directly in the module files. Either way, it might be best to separate type-hinting the Python API (i.e. covering all of the arguments and outputs of public methods) from a full type-hinting effort (i.e. covering internals and passing all static type analyses), where the latter is where I expect most of the tricky Cython-specific syntax to be required.

**Comment by bryanwweber** [comment]
Thanks @TimothyEDawson ! I agree with @speth that directly integrated hints to start with would be preferable. I think type hints in general have proven their utility and I think it's worth having official support from the start. 

I also agree with your approach to provide type hints for the public API as the first priority. We can experiment with the location for the hints, but I suspect that separate pyi files will help avoid having to provide hints for the internal details. For example, there may be some cases in the actual source where we do complicated or dynamic behaviors to generate a much simpler public API. Hinting that in the source would be a pain, but using pyi files allows us to just document the final interface.

**Comment by TimothyEDawson** [comment]
@bryanwweber Good point, I started working on this yesterday and I had underestimated how much dynamic stuff was happening under the hood! Several things which could cause headaches for type hinting, such as the various runtime-generated properties + methods for classes like `Quantity` and `SolutionArray`.

I'm making quick progress on creating .pyi files which can comfortably live in `interfaces/cython/cantera`. Once I have some time, I'll open a draft pull request and link to it here!

**Comment by TimothyEDawson** [comment]
Finally got around to it, I look forward to everyone's feedback and any contributions!

https://github.com/Cantera/cantera/pull/1926
